File: player.py
"""AI玩家 - 通过 litellm 调用 LLM"""
from __future__ import annotations
import json
import logging
from typing import Dict, Any, Optional
from litellm import completion
from models import PlayerState, Role

logger = logging.getLogger("ai-werewolf")


class AIPlayer:
    """封装 LLM 调用的玩家"""

    def __init__(self, state: PlayerState, max_retries: int = 3,
                 temperature: float = 0.7, max_tokens: int = 512,
                 dry_run: bool = False):
        self.state = state
        self.max_retries = max_retries
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.dry_run = dry_run

    @property
    def id(self) -> int:
        return self.state.player_id

    @property
    def name(self) -> str:
        return self.state.name

    @property
    def role(self) -> Role:
        return self.state.role

    @property
    def alive(self) -> bool:
        return self.state.alive

    def chat(self, system_prompt: str, user_prompt: str,
             verbose: bool = False) -> str:
        """调用 LLM，返回原始文本响应"""
        if self.dry_run:
            return self._dry_run_response(system_prompt, user_prompt)

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

        for attempt in range(1, self.max_retries + 1):
            try:
                resp = completion(
                    model=self.state.model,
                    messages=messages,
                    temperature=self.temperature,
                    max_tokens=self.max_tokens,
                    timeout=60,
                )
                text = resp.choices[0].message.content.strip()
                if verbose:
                    logger.info(f"[{self.name}] LLM响应 (attempt {attempt}): {text[:200]}")
                return text
            except Exception as e:
                logger.warning(
                    f"[{self.name}] LLM调用失败 (attempt {attempt}/{self.max_retries}): {e}"
                )
                if attempt == self.max_retries:
                    logger.error(f"[{self.name}] LLM调用最终失败，返回空响应")
                    return ""
        return ""

    def chat_json(self, system_prompt: str, user_prompt: str,
                  verbose: bool = False) -> Optional[Dict[str, Any]]:
        """调用 LLM 并期望返回 JSON"""
        text = self.chat(system_prompt, user_prompt, verbose)
        if not text:
            return None
        return text  # 解析交给 parser

    def _dry_run_response(self, system_prompt: str, user_prompt: str) -> str:
        """dry-run 模式下返回模拟响应"""
        import random

        # 提取可选的合法ID
        valid_ids = []
        for m in __import__('re').finditer(r'\[(\d+)\]', user_prompt):
            valid_ids.append(int(m.group(1)))

        # 优先判断发言和投票（这些prompt也可能包含"狼人"等关键词）
        if "发言" in user_prompt or "speech" in user_prompt.lower():
            speeches = [
                "我觉得昨晚的死讯很蹊跷，需要仔细分析一下每个人的表现。",
                "我是好人，请大家相信我。我建议大家关注一下发言异常的人。",
                "有没有人要跳预言家？我觉得现在应该分享一些信息了。",
                "我观察了一下，有些人的发言比较模糊，可能是在隐藏什么。",
                "昨晚的死亡顺序值得深思，我觉得狼人可能在故意引导方向。",
                "根据我的分析，目前局势比较复杂，我们需要谨慎投票。",
            ]
            return json.dumps({"speech": random.choice(speeches)}, ensure_ascii=False)
        elif "投票" in user_prompt or "vote" in user_prompt.lower():
            # 排除自己
            my_id_match = __import__('re').search(r'编号\s*(\d+)', user_prompt)
            my_id = int(my_id_match.group(1)) if my_id_match else 0
            choices = [x for x in valid_ids if x != my_id] or [1]
            return json.dumps({"vote": random.choice(choices)}, ensure_ascii=False)
        elif "狼人" in user_prompt and "杀" in user_prompt:
            choices = valid_ids or [1]
            return json.dumps({"kill": random.choice(choices)}, ensure_ascii=False)
        elif "查验" in user_prompt:
            choices = valid_ids or [2]
            return json.dumps({"check": random.choice(choices)}, ensure_ascii=False)
        elif "女巫" in user_prompt:
            return json.dumps({"save": random.random() > 0.5, "poison": None}, ensure_ascii=False)
        elif "猎人" in user_prompt and ("开枪" in user_prompt or "射击" in user_prompt):
            choices = valid_ids or [1]
            return json.dumps({"shoot": random.choice(choices)}, ensure_ascii=False)
        else:
            return json.dumps({"action": "pass"}, ensure_ascii=False)

    def __repr__(self) -> str:
        status = "存活" if self.alive else "死亡"
        return f"Player({self.id}|{self.name}|{self.role.display_name}|{status})"

File: lib/llm/player.ts
/**
 * AIPlayer —— Vercel AI SDK 封装，对应原 Python 版 player.py:AIPlayer
 *
 * 提供：
 *   - chat(): 调用 LLM 返回原始文本
 *   - dry-run 模式：返回模拟响应，不消耗 token
 *   - 重试：失败时按指数退避重试
 *   - 指标采集：记录 tokens 用量、响应耗时（供后续统计）
 */
import { generateText, type LanguageModel } from 'ai';
import type { PlayerState, LlmUsage } from '../types';
import { getModel } from './providers';
import { dryRunResponse } from './dryRun';

export interface ChatResult {
  text: string;
  usage: LlmUsage | null;
}

/**
 * 从 AI SDK 抛出的错误里抽出最有用的诊断字段：
 *   - url           → 实际请求的端点（检查 baseURL 是否带 /v1）
 *   - statusCode    → 401/404/500 等
 *   - responseBody  → 服务端返回的原始内容（前 500 字符）
 *
 * Vercel AI SDK v6 的 APICallError 暴露这些字段，但默认 err.message 经常只说
 * "Invalid JSON response" 这种没用的信息——用户看不到真正的 HTTP 状态。
 */
function describeError(err: Error | null): { brief: string; detail: string } {
  if (!err) return { brief: 'unknown', detail: '' };
  const brief = err.message || err.name || 'unknown';
  // APICallError 字段（不强类型依赖，duck-typing 即可）
  const anyErr = err as Error & {
    url?: string;
    statusCode?: number;
    responseBody?: string;
    responseHeaders?: Record<string, string>;
    requestBodyValues?: unknown;
  };
  const lines: string[] = [];
  if (anyErr.url) lines.push(`  url: ${anyErr.url}`);
  if (anyErr.statusCode !== undefined) lines.push(`  status: ${anyErr.statusCode}`);
  const body = anyErr.responseBody?.trim();
  if (body) {
    lines.push(`  body[0..500]: ${body.slice(0, 500)}`);
  }
  if (!lines.length) return { brief, detail: `  ${err.name}: ${err.message}` };
  return { brief, detail: lines.join('\n') };
}

export interface AIPlayerConfig {
  maxRetries?: number;
  temperature?: number;
  maxTokens?: number;
  dryRun?: boolean;
  verbose?: boolean;
}

export class AIPlayer {
  readonly state: PlayerState;
  readonly maxRetries: number;
  readonly temperature: number;
  readonly maxTokens: number;
  readonly dryRun: boolean;
  readonly verbose: boolean;
  /** 最近一次调用的指标，phase 层读取后清零 */
  lastUsage: LlmUsage | null = null;

  constructor(state: PlayerState, config: AIPlayerConfig = {}) {
    this.state = state;
    this.maxRetries = config.maxRetries ?? 3;
    this.temperature = config.temperature ?? 0.7;
    this.maxTokens = config.maxTokens ?? 512;
    this.dryRun = config.dryRun ?? false;
    this.verbose = config.verbose ?? false;
  }

  get id(): number {
    return this.state.playerId;
  }

  get name(): string {
    return this.state.name;
  }

  get role() {
    return this.state.role;
  }

  get alive(): boolean {
    return this.state.alive;
  }

  async chat(systemPrompt: string, userPrompt: string): Promise<string> {
    const result = await this.chatWithUsage(systemPrompt, userPrompt);
    return result.text;
  }

  async chatWithUsage(systemPrompt: string, userPrompt: string): Promise<ChatResult> {
    if (this.dryRun) {
      const text = dryRunResponse(userPrompt);
      const usage: LlmUsage = {
        promptTokens: 0,
        completionTokens: 0,
        totalTokens: 0,
        durationMs: 0,
        model: `${this.state.model} (dry-run)`,
      };
      this.lastUsage = usage;
      return { text, usage };
    }

    let model: LanguageModel;
    try {
      model = getModel(this.state.model, {
        baseURL: this.state.baseUrl,
        apiKey: this.state.apiKey,
      });
    } catch (e) {
      console.error(`[${this.name}] 无法解析模型 ${this.state.model}: ${(e as Error).message}`);
      this.lastUsage = null;
      return { text: '', usage: null };
    }

    let lastError: Error | null = null;
    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      const start = Date.now();
      try {
        const { text, usage } = await generateText({
          model,
          system: systemPrompt,
          messages: [{ role: 'user', content: userPrompt }],
          temperature: this.temperature,
          maxOutputTokens: this.maxTokens,
        });
        const durationMs = Date.now() - start;

        const promptTokens = usage?.inputTokens ?? 0;
        const completionTokens = usage?.outputTokens ?? 0;
        const llmUsage: LlmUsage = {
          promptTokens,
          completionTokens,
          totalTokens: promptTokens + completionTokens,
          durationMs,
          model: this.state.model,
        };
        this.lastUsage = llmUsage;

        if (this.verbose) {
          console.info(
            `[${this.name}] LLM 响应 (attempt ${attempt}, ${durationMs}ms, ${llmUsage.totalTokens} tokens): ${text.slice(0, 200)}`,
          );
        }
        return { text, usage: llmUsage };
      } catch (e) {
        const err = e as Error;
        lastError = err;
        const diag = describeError(err);
        if (attempt === 1) {
          // 第一次失败时打印完整诊断（URL/状态/响应体），后续重试只打印简略
          console.warn(
            `[${this.name}] LLM 调用失败 (attempt ${attempt}/${this.maxRetries}): ${diag.brief}\n${diag.detail}`,
          );
        } else {
          console.warn(
            `[${this.name}] LLM 调用失败 (attempt ${attempt}/${this.maxRetries}): ${diag.brief}`,
          );
        }
        // 指数退避
        if (attempt < this.maxRetries) {
          const delay = 500 * 2 ** (attempt - 1);
          await new Promise((r) => setTimeout(r, delay));
        }
      }
    }

    const finalDiag = describeError(lastError);
    console.error(`[${this.name}] LLM 调用最终失败: ${finalDiag.brief}\n${finalDiag.detail}`);
    this.lastUsage = null;
    return { text: '', usage: null };
  }

  toString(): string {
    const status = this.alive ? '存活' : '死亡';
    const role = this.state.role;
    return `Player(${this.id}|${this.name}|${role}|${status})`;
  }
}
